ergo-mempool-sync/src/lib.rs

```diff
@@ -158,49 +114,80 @@ async fn sync<'a, TClient: ErgoNetwork>(client: &TClient, mut state: RefMut<'a,
+                error,
+            );
+            START_HEIGHT
+        }
+    };
+    let chain_sync = chain_sync_maker.init(start_at as u32, None).await;
+    let state = Arc::new(Mutex::new(SyncState::empty()));
+    let joined_stream = select_all(vec![
+        boxed(sync_ledger(chain_sync_stream(chain_sync), Arc::clone(&state)).map(move |_| None)),
+        boxed(sync_mempool(conf, client, state)),
+    ]);
+    joined_stream.filter_map(move |maybe_upd| async move { maybe_upd })
+}
+
+fn sync_ledger<'a, S>(upstream: S, state: Arc<Mutex<SyncState>>) -> impl Stream<Item = ()> + 'a
+where
+    S: Stream<Item = ChainUpgrade> + 'a,
 {
-    type Item = MempoolUpdate;
-
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        let mut this = self.project();
-        let client: &TClient = this.client;
-        if Future::poll(Pin::new(&mut this.conf.sync_interval), cx).is_ready() {
-            let mut state = this.state.borrow_mut();
-            // Sync chain tail
-            loop {
-                match Stream::poll_next(this.chain_sync.as_mut(), cx) {
-                    Poll::Ready(Some(ChainUpgrade::RollForward(blk))) => state.push_block(blk),
-                    Poll::Ready(Some(ChainUpgrade::RollBackward(_))) => state.pop_block(),
-                    Poll::Ready(None) => {}
-                    Poll::Pending => break,
+    upstream.then(move |upgrade| {
+        let state = Arc::clone(&state);
+        async move {
+            let mut state = state.lock().await;
+            match upgrade {
+                ChainUpgrade::RollForward(blk) => {
+                    state.push_block(blk);
                 }
-            }
-            let mut sync_fut = Box::pin(sync(client, state));
-            // Drive sync to completion
-            loop {
-                match sync_fut.as_mut().poll(cx) {
-                    Poll::Ready(_) => break,
-                    Poll::Pending => continue,
+                ChainUpgrade::RollBackward(_) => {
+                    state.pop_block();
                 }
             }
         }
-        if let Some(upgr) = this.state.borrow_mut().pending_updates.pop_front() {
-            return Poll::Ready(Some(upgr));
-        }
-        Poll::Pending
-    }
+    })
 }
 
-impl<'a, TClient, TChainSync> FusedStream for MempoolSync<'a, TClient, TChainSync>
+fn sync_mempool<'a, TClient>(
+    conf: MempoolSyncConf,
+    client: &'a TClient,
+    state: Arc<Mutex<SyncState>>,
+) -> impl Stream<Item = Option<MempoolUpdate>> + 'a
 where
-    MempoolSync<'a, TClient, TChainSync>: Stream,
+    TClient: ErgoNetwork,
 {
-    /// MempoolSync stream is never terminated.
-    fn is_terminated(&self) -> bool {
-        false
+    stream! {
+        loop {
+            let mut st = state.lock().await;
+            if let Some(upd) = st.pending_updates.pop_front() { // First, try to pop updates
+                yield Some(upd)
+            } else { // Wait otherwise
```

Comment:
```suggestion
            } else { // Wait otherwise
                drop(st);
```
We should drop the lock before we delay.